api/v1/cluster_webhook.go

```diff
@@ -799,6 +800,56 @@ func (r *Cluster) validateBootstrapRecoverySource() field.ErrorList {
 	return result
 }
 
+// validateBootstrapRecoveryDataSource is used to ensure that the data
+// source is correctly defined
+func (r *Cluster) validateBootstrapRecoveryDataSource() field.ErrorList {
+	var result field.ErrorList
+
+	// This validation is only applicable for datasource-based recovery based bootstrap
+	if r.Spec.Bootstrap == nil || r.Spec.Bootstrap.Recovery == nil || r.Spec.Bootstrap.Recovery.DataSource == nil {
+		return result
+	}
+
+	recoveryStanza := r.Spec.Bootstrap.Recovery
+	if recoveryStanza.Source != "" {
+		result = append(
+			result,
+			field.Invalid(
+				field.NewPath("spec", "bootstrap", "recovery", "dataSource"),
+				r.Spec.Bootstrap.Recovery.DataSource,
+				"Recovery from dataSource is not compatible with other types of recovery"))
+	}
+
+	if recoveryStanza.Backup != nil {
+		result = append(
+			result,
+			field.Invalid(
+				field.NewPath("spec", "bootstrap", "recovery", "backup"),
+				r.Spec.Bootstrap.Recovery.Backup,
+				"Recovery from dataSource is not compatible with other types of recovery"))
+	}
+
+	if recoveryStanza.RecoveryTarget != nil {
+		result = append(
+			result,
+			field.Invalid(
+				field.NewPath("spec", "bootstrap", "recovery", "recoveryTarget"),
+				r.Spec.Bootstrap.Recovery.RecoveryTarget,
+				"A recovery target cannot be set while recoverying from a DataSource"))
+	}
+
+	if recoveryStanza.DataSource.WalStorage != nil && r.Spec.WalStorage == nil {
+		result = append(
+			result,
+			field.Invalid(
+				field.NewPath("spec", "bootstrap", "recovery", "dataSource", "walStorage"),
+				r.Spec.Bootstrap.Recovery.DataSource.WalStorage,
+				"A WAL storage configuration is required when recoverying using a DataSource for WALs"))
+	}
+
+	return result
```

Comment:
Should we check here also if the sizes of the storages are larger than the ones of the snapshots? The pod will remain pending and only the PVC will display the cause of the issue.

Reply:
We could, but to do that we need the webhook server to reach the API server. And since the webhook server is handing a request created by the API server itself... I think we could stress the API server too much.

Beside this, in my understanding the VolumeSnapshot specification don't actually enforce this constraint, which is enforced by storage implementations. There could be storage implementations allowing changing the size on the fly, and perhaps we should not prevent that.